aten/src/ATen/cuda/CUDAGenerator.cpp

```diff
@@ -1,46 +1,148 @@
+  });
+  return default_gens_cuda[idx].get();
 }
 
-CUDAGenerator& CUDAGenerator::copy(const Generator& from) {
-  throw std::runtime_error("CUDAGenerator::copy() not implemented");
+/**
+ * Utility to create a CUDAGenerator. Returns a shared_ptr
+ */
+std::shared_ptr<CUDAGenerator> createCUDAGenerator(DeviceIndex device_index) {
+  std::call_once(num_gpu_init_flag, initCUDAGenVector);
+  DeviceIndex idx = device_index;
+  if (idx == -1) {
+    idx = c10::cuda::current_device();
+  }
+  TORCH_CHECK(idx >= 0 && idx < num_gpus, "The device_index is invalid.");
+  auto gen = std::make_shared<CUDAGenerator>(idx);
+  gen->set_current_seed(default_rng_seed_val);
+  gen->set_philox_offset_per_thread(0);
+  return gen;
 }
 
-CUDAGenerator& CUDAGenerator::free() {
-  THCRandom_shutdown(context->getTHCState());
-  return *this;
+} // namespace detail
+} // namespace cuda
+
+/**
+ * CUDAGenerator class implementation
+ */
+CUDAGenerator::CUDAGenerator(DeviceIndex device_index)
+  : Generator{Device(DeviceType::CUDA, device_index)} { }
+
+/**
+ * Sets the seed to be used by curandStatePhilox4_32_10
+ * Resets the philox_offset_per_thread_ to 0
+ * 
+ * See Note [Acquire lock when using random generators]
+ */
+void CUDAGenerator::set_current_seed(uint64_t seed) {
+  seed_ = seed;
+  philox_offset_per_thread_ = 0;
+}
+
+/**
+ * Gets the current seed of CUDAGenerator.
+ */
+uint64_t CUDAGenerator::current_seed() const {
+  return seed_;
 }
 
+/**
+ * Gets a nondeterminstic random number from /dev/urandom or time,
+ * seeds the CPUGenerator with it and then returns that number.
+ * 
+ * FIXME: You can move this function to Generator.cpp if the algorithm
+ * in getNonDeterministicRandom is unified for both CPU and CUDA
+ */
 uint64_t CUDAGenerator::seed() {
-  return THCRandom_initialSeed(context->getTHCState());
+  auto random = detail::getNonDeterministicRandom(true);
+  this->set_current_seed(random);
+  return random;
 }
 
-uint64_t CUDAGenerator::current_seed() const {
-  return THCRandom_initialSeed(context->getTHCState());
+/**
+ * Sets the philox_offset_per_thread_ to be used by curandStatePhilox4_32_10
+ * 
+ * See Note [Acquire lock when using random generators]
+ */
+void CUDAGenerator::set_philox_offset_per_thread(uint64_t offset) {
+  philox_offset_per_thread_ = offset;
 }
 
-void CUDAGenerator::set_current_seed(uint64_t seed) {
-  THCRandom_manualSeed(context->getTHCState(), seed);
+/**
+ * Gets the current philox_offset_per_thread_ of CUDAGenerator.
+ */
+uint64_t CUDAGenerator::philox_offset_per_thread() {
+  return philox_offset_per_thread_;
+}
+
+/**
+ * Gets the seed and philox offset value to be used in
+ * curandStatePhilox4_32_10
+ * 
+ * Each kernel using philox has to sensibly increment offset
+ * for future users of philox. So it gets the "old" value for
+ * itself (before add), and tells subsequent users which offset
+ * they should use, since only the kernel knows how many randoms
+ * it intends to generate. 
```

Comment:
I can't wait for us to switch CPU to Philox; then we could probably get rid of locking entirely, right? In fact, we can probably make this CUDA code lock free by making just `this->philox_offset_per_thread_ += increment;` atomic

Reply:
We need someone to vectorize the Philox header with the fancy AVX vector utilities in PyTorch :). In my attempt to switch the CPU rng engine to Philox, I found that the single thread performance of Philox is about 1.5X slower that mt19937. But yes, that is something for the future - may be thread local splittable rngs :) ? 

The `philox_offset_per_thread` was atomic and was updated atomically using `next_philox_seed`, but I made it non-atomic since I moved it to being a member function of CUDAGenerator and and there our policy is to use lock_guard explicitly.

Reply:
Hi there, I've been studying the current (2021) implementation of random generators in Pytorch, and it seems there is still no splittable API. Is this something people are still interested in?

Reply:
I believe you can have this discussion with the current maintainers in the #rng channel of pytorch slack :).